contracts/cw721-base/src/contract.rs

```diff
@@ -14,442 +12,609 @@ use cw721::{
 
-    if info.sender != minter {
-        return Err(ContractError::Unauthorized {});
+    pub fn num_tokens(&self, storage: &dyn Storage) -> StdResult<u64> {
+        Ok(self.token_count.may_load(storage)?.unwrap_or_default())
     }
 
-    // create the token
-    let token = TokenInfo {
-        owner: deps.api.addr_validate(&msg.owner)?,
-        approvals: vec![],
-        name: msg.name,
-        description: msg.description.unwrap_or_default(),
-        image: msg.image,
-    };
-    tokens().update(deps.storage, &msg.token_id, |old| match old {
-        Some(_) => Err(ContractError::Claimed {}),
-        None => Ok(token),
-    })?;
-
-    increment_tokens(deps.storage)?;
-
-    Ok(Response::new()
-        .add_attribute("action", "mint")
-        .add_attribute("minter", info.sender)
-        .add_attribute("token_id", msg.token_id))
+    pub fn increment_tokens(&self, storage: &mut dyn Storage) -> StdResult<u64> {
+        let val = self.num_tokens(storage)? + 1;
+        self.token_count.save(storage, &val)?;
+        Ok(val)
+    }
 }
 
-pub fn execute_transfer_nft(
-    deps: DepsMut,
-    env: Env,
-    info: MessageInfo,
-    recipient: String,
-    token_id: String,
-) -> Result<Response, ContractError> {
-    _transfer_nft(deps, &env, &info, &recipient, &token_id)?;
-
-    Ok(Response::new()
-        .add_attribute("action", "transfer_nft")
-        .add_attribute("sender", info.sender)
-        .add_attribute("recipient", recipient)
-        .add_attribute("token_id", token_id))
-}
+impl<'a, T, C> Cw721Contract<'a, T, C>
+where
+    T: Serialize + DeserializeOwned + Clone,
+    C: Clone + std::fmt::Debug + PartialEq + JsonSchema,
+{
+    pub fn instantiate(
+        &self,
+        deps: DepsMut,
+        _env: Env,
+        _info: MessageInfo,
+        msg: InstantiateMsg,
+    ) -> StdResult<Response<C>> {
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+        let info = ContractInfoResponse {
+            name: msg.name,
+            symbol: msg.symbol,
+        };
+        self.contract_info.save(deps.storage, &info)?;
+        let minter = deps.api.addr_validate(&msg.minter)?;
+        self.minter.save(deps.storage, &minter)?;
+        Ok(Response::default())
+    }
 
-pub fn execute_send_nft(
-    deps: DepsMut,
-    env: Env,
-    info: MessageInfo,
-    contract: String,
-    token_id: String,
-    msg: Binary,
-) -> Result<Response, ContractError> {
-    // Transfer token
-    _transfer_nft(deps, &env, &info, &contract, &token_id)?;
-
-    let send = Cw721ReceiveMsg {
-        sender: info.sender.to_string(),
-        token_id: token_id.clone(),
-        msg,
-    };
-
-    // Send message
-    Ok(Response::new()
-        .add_message(send.into_cosmos_msg(contract.clone())?)
-        .add_attribute("action", "send_nft")
-        .add_attribute("sender", info.sender)
-        .add_attribute("recipient", contract)
-        .add_attribute("token_id", token_id))
-}
+    pub fn execute(
```

Comment:
I would definitely move this out of this struct implementation, probably to the structure message, and make it dispatching over any contract which implements a trait representing the interface of this contract. In this particular case, the trait would be a `Cw721`, and the `execute` would be:

```rust
impl<T> ExecMsg<T> {
  /// Dispatches this message to contract
  fn execute<Contract: Cw721>(self, contract: &Contract) -> Result<Response<Contract::C>, Contract::Err> {
    match self {
      ExecuteMsg::Mint(msg) => self.mint(deps, env, info, msg),
      // All other branches
    }
  }
}
```

`Err` and `C` (Which should get some meaningfull name - there is no restriction on generic arguments identifier lengths ;) ) are dependent on actual contract implementation (as it may and possibly would differ for different contracts).

Reply:
This makes sense.

The problem is this is also not reusable if they add even one field to the ExecMsg in some extension. They would need something like `impl TryFrom<CustomExecMsg> for ExecMsg`, which would be quite elegant if we could use [`#[serde(flatten)]`](https://serde.rs/attr-flatten.html) but we [cannot do that yet](https://github.com/CosmWasm/serde-json-wasm/issues/20).

I could do this inversion of control (no problem), but it is only really useful if all derived contracts have the exact same message structure or we can use flatten to extend them.

Reply:
It still can be done bypassing serde (implementing `Serialize` just by forwarding calls to the enums), but it looks like much to be done (except if it may be generated in which case it is free).